File: core/shared/src/main/scala/diplomacy/variant/standard/rule/Validator.scala
package diplomacy.variant.standard.rule

import diplomacy.rule.InvalidOrderMessage
import diplomacy.board.Power

trait Validator[Turn_ <: Turn, Power_ <: Power] extends Rule.TypeHelper {
  final type Turn = Turn_
  final type Power = Power_

  def unitsRequiringOrder(board: Board): Set[DiplomacyUnit]

  def errorMessageOfOrder(board: Board)(order: Order): Option[InvalidOrderMessage]
  def errorMessageOfOrders(board: Board)(order: Set[Order]): Option[InvalidOrderMessage]
}

File: src/variant/standard/rule/validator.ts
import { Unit, Board, ResolvedResult } from "./types"
import { Order } from "./order"
import { Error } from "./error"

export interface Validator<Power> {
  /**
   * @return The set of units that are requred orders.
   */
  unitsRequiringOrder (board: Board<Power>): Set<Unit<Power>>
  /**
   * @return The error message of the order. If the order is valid, it's null.
   */
  errorMessageOfOrder (board: Board<Power>, order: Order<Power>): Error | null
  /**
   * @return The error message of the orders. If the set of the orders is valid, it's null.
   */
  errorMessageOfOrders (board: Board<Power>, orders: Set<Order<Power>>): Error | null
}
